nfc: add basic Mifare DESFire support

kevinwallace opened the pull request:
# What's new

- Detection of DESFire cards based on ATQA/SAK
- Reading of all non-protected card data
- UI to browse apps/files in card data
- Fixes to TextBox wrap behavior
- Adds `bool` type to flipper_format
- Saving and loading DESFire tag data

# Verification 

- Open NFC app and scan a DESFire card.  Here is an example of me scanning an ORCA (Seattle public transit) card and briefly browsing its data.

![flipper_desfire_orca](https://user-images.githubusercontent.com/184975/158650853-35ca3701-957b-4005-b71c-aa9413f08747.jpg)

https://user-images.githubusercontent.com/184975/158492956-21d614c1-8e55-452a-bb48-48a1f1f53f81.mov

# Checklist (do not modify)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

Commit:
Fix TextBox word wrap behavior

* Wrap width is 120 pixels, not 140. (140 is larger than the screen!)
* Glyph width already includes spacing; don't add 1 additional px
* When starting a new line, include wrapped glyph width in new line_width.
* Call canvas_set_font before text_box_insert_endline so that glyph
  width is calculated using correct font.

Previous approach worked somewhat well using default TextBoxFontText but
this version is more robust, particularly when using TextBoxFontHex.

Commit:
Add basic Mifare DESFire reading, file/app browser

Commit:
Fix build with APP_ARCHIVE=0

/project/applications/desktop/scenes/desktop_scene_main.c:51:13: error: 'desktop_switch_to_app' defined but not used [-Werror=unused-function]
   51 | static void desktop_switch_to_app(Desktop* desktop, const FlipperApplication* flipper_app) {
      |             ^~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [/project/make/rules.mk:70: .obj/f7//project/applications/desktop/scenes/desktop_scene_main.o] Error 1

(Discovered when turning off all the other APP_*s to free up enough space to use APP_UNIT_TESTS)

Commit:
Add bool type to flipper_format

Built with:
  make APP_UNIT_TESTS=1 SRV_DESKTOP=0 APP_ARCHIVE=0 APP_GPIO=0 APP_IBUTTON=0 APP_INFRARED=0 APP_LF_RFID=0 APP_NFC=0 APP_SUBGHZ=0 APP_ABOUT=0 APP_PASSPORT=0 APP_MUSIC_PLAYER=0 APP_SNAKE_GAME=0 APP_ACCESSOR=0 APP_BLINK=0 APP_INFRARED_MONITOR=0 APP_KEYPAD_TEST=0 APP_SD_TEST=0 APP_VIBRO_TEST=0 APP_USB_TEST=0 APP_DISPLAY_TEST=0 APP_BLE_HID=0 APP_USB_MOUSE=0 APP_BAD_USB=0 APP_U2F=0 APP_UART_ECHO=0
and passed `unit_tests` via cli

Commit:
Add ability to save and load DESFire card data

kevinwallace commented:
Updated to include saving/loading of DESFire tag data.  Let me know if the flipper_format bool change is unwanted -- I'm happy to drop that commit and rewrite to use hex for those fields instead.

skotopes commented:
Hi @kevinwallace 

Great amazing job, you are awesome! )

Adding bool is ok. @gornekich will review PR and provide feedback.

Commit:
Skip over NfcSceneDeviceInfo when viewing saved DESFire info

All of the same information is present in NfcSceneReadMifareDesfireSuccess,
except for ATQA/SAK, which is implicitly known for all saved DESFire cards
(0x4403, 0x20).

kevinwallace commented:
UI has changed a bit since the original video, here's the current state of things

https://user-images.githubusercontent.com/184975/158708890-6bd59e9b-79b8-4e40-94dd-44851dc0bdf6.mov

(and thank **you** for Flipper -- this has been such a fun device to hack on!  :)

skotopes committed:
Merge branch 'dev' into desfire

Commit:
mf_df_clear: don't leak master key settings key versions

gornekich commented on applications/nfc/nfc.c:
```diff
@@ -163,6 +163,8 @@ int32_t nfc_app(void* p) {
         if(nfc->dev->format == NfcDeviceSaveFormatMifareUl) {
             scene_manager_next_scene(nfc->scene_manager, NfcSceneEmulateMifareUl);
+        } else if(nfc->dev->format == NfcDeviceSaveFormatMifareDesfire) {
+            scene_manager_next_scene(nfc->scene_manager, NfcSceneReadMifareDesfireSuccess);
```
Now we start nfc application with parameters if we want to launch NFC app from Archive and start UID or Mifare Ultralight emulation immediately. In your case, we will see information about selected DESFire card. I am not sure that we want to break user's expectation here, so I would delete this for now

kevinwallace replied:
Done!  8b85ff3f854976834b8974ca390c51ce05d47f9d

gornekich commented on applications/nfc/nfc_worker.c:
```diff
@@ -658,6 +670,242 @@ void nfc_worker_emulate_mifare_ul(NfcWorker* nfc_worker) {
+                    continue;
+                }
+                if(!mf_df_parse_read_data_response(rx_buff, rx_len, file)) {
+                    FURI_LOG_W(TAG, "Bad response reading file %d\r\n", file->id);
```
No need in \r\n here

kevinwallace replied:
Oops!  This was originally a printf.  Done.  ff1fcbf55f11f6f88d93b84f9d6158491d25536a

gornekich commented on applications/nfc/scenes/nfc_scene_mifare_desfire_app.c:
```diff
@@ -0,0 +1,109 @@
+    MifareDesfireApplication* app = nfc_scene_mifare_desfire_app_get_app(nfc);
+    if(!app) {
+        FURI_LOG_E(TAG, "Bad state. No app selected?");
+        return;
```
There is a risk of undesired behavior. If you return here, view dispatcher will display and deliver events to the view from previous scene and execute on_event() handler of MifareDesfireApp scene. I suggest switching to a Popup view with error explanation.

kevinwallace replied:
If this codepath ever runs, it's a bug in the MifareDesfireData scene.  But just in case I made a nice popup anyway.  :)  f5dbd6fd781180897596584a32b5dbacb6fd6de9

![error_popup](https://user-images.githubusercontent.com/184975/158894894-d8396307-eafb-4a37-be86-43ec3d7480bd.png)

gornekich reviewed (changes requested):
Hello @kevinwallace !
Thank you very much for your amazing work!
So far, I looked through your code and left a few comments. I need more time to dive into your DESFire implementation more carefully, and make detailed review.
I noticed a memory leakage after I open DESFire card and then exit NFC application. You can see this from our log console (LoaderSrv logs) or executing 'free' CLI command before and after launching NFC app and opening DESFire card. Memory leakage is a blocking issue for merging your pull request.
I suggest you fix memory issues while I will make detailed review in next couple of days.

Commit:
When opening a DESFire card from Archive, retain UID emulation behavior

Commit:
rm unnecessary \r\n

Commit:
show Popup instead of leaving view in bad state

kevinwallace commented:
Thank you for the comments!

A few hours ago, I noticed a memory leak while reviewing this PR's diff and fixed it in 56ab13eab0703e88ea7b7e83ff61cf10cca20a5f.  I think this was the same leak you saw -- I'm now checking with the `free` CLI command and `Free heap size` is remaining at the same size whenever I return to the desktop from the NFC app.  Could you let me know if you're still seeing this, and if there's a particular set of interactions to reproduce it?

gornekich commented on applications/nfc/scenes/nfc_scene_saved_menu.c:
```diff
@@ -62,7 +74,11 @@ bool nfc_scene_saved_menu_on_event(void* context, SceneManagerEvent event) {
         } else if(event.event == SubmenuIndexInfo) {
-            scene_manager_next_scene(nfc->scene_manager, NfcSceneDeviceInfo);
+            if(nfc->dev->format == NfcDeviceSaveFormatMifareDesfire) {
+                scene_manager_next_scene(nfc->scene_manager, NfcSceneReadMifareDesfireSuccess);
```
I think it's better to show DESFire cards Info in NfcSceneDeviceInfo, or make another NfcSceneMifareDesfireInfo scene. If I go to  NfcSceneReadMifareDesfireSuccess scene from **Saved cards**, I don't expect to **Name and save** card that I already opened

kevinwallace replied:
Agreed that offering "Name and save" for an already-saved card is inappropriate.  I've added support for DESFire cards to `NfcSceneDeviceInfo` and now go there instead of `NfcSceneReadMifareDesfireSuccess` when reading a saved card.  The "Data" button goes to `NfcSceneMifareDesfireData`, and `NfcSceneMifareDesfireMenu` is now only accessible when scanning a new card.

bf643267394ed3ee8988eabc2af02c64d5b6aee4

![device_info](https://user-images.githubusercontent.com/184975/159061589-ebb4c878-1674-4da0-a6bb-c67fb9cf80a9.png)

I also tried another approach, keeping `NfcSceneReadMifareDesfireSuccess` in both cases, and showing the 'More' button selectively based on scene history.  It's in my `desfire_wip` branch (https://github.com/kevinwallace/flipperzero-firmware/commit/26ef6e03d5c82c66c64935c6e0665f923eb252b0, https://github.com/kevinwallace/flipperzero-firmware/commit/a42c938b6e27c3c952836fbeb20f93e10dab9fc8) if you'd like to take a look, but after trying both, I prefer the approach in this PR.

gornekich replied:
I like current approach with adding DESFire to NfcSceneDeviceInfo scene

gornekich reviewed (commented):
Perfect! The leakage I mentioned is gone. Also, thanks for following my suggestions!

Another issue I found is that if I go to saved DESFire card -> Emulate UID -> Back -> Info, there is no info about saved card and after I exit NFC ~3kB of heap leak.
Current UID emulation supports displaying reader commands on the screen. The reason behind this is that we give users an opportunity to research what kind of reader in front of them. From technical point of view, we pass commands received in **nfc worker** to **emulate uid scene** through NfcReaderRequestData structure in NfcDeviceData, which shares memory with MifareDesfireData. That's where after UID emulation we loose DESFire data and can't free allocated memory.
The quickest fix here is to disable UID emulation for DESFire cards, just like it's done for Bank Cards. However, if you want to keep opportunity for UID emulation, you should fix NfcDeviceData structure  so that NfcReaderRequestData and MifareDesfireData don't share same memory.

Let me know how you want to fix UID emulation issue and when it's done, we will merge your brilliant PR :)

Commit:
Move NfcReaderRequestData out of union

This makes it safe to emulate DESFire/EMV without clobbering card data.

Commit:
Display saved DESFire cards via NfcSceneDeviceInfo

kevinwallace commented:
Good catch on the broken UID emulation behavior due to memory sharing -- I didn't look closely at the UID emulation code earlier to understand this was going on.  It seems useful to still be able to emulate UID only, so I've moved `NfcReaderRequestData` out of the union so that it no longer shares memory with DESFire data.  This means it's possible to safely emulate UID of EMV cards from the saved card menu, too, so I've also re-enabled it for those.

fcc1bfa728c716a12e948819df9edea033848cd6

I'm also fine disabling it and leaving `NfcReaderRequestData` in the union if you prefer that approach -- let me know and I'll revise.

Commit:
Display and save file metadata even when contents are missing

This can happen when a file doesn't allow unauthenticated reads (see the
call to mf_df_parse_read_data_response in nfc_worker.c).

gornekich committed:
Merge branch 'dev' into desfire

gornekich commented:
Thanks for fixing UID emulation. Now everything looks good!

gornekich committed:
Merge branch 'dev' into desfire

skotopes committed:
Merge branch 'dev' into desfire

gornekich commented:
Thanks for contribution!
